ITfoxtec.Saml2/Bindings/Saml2RedirectBinding.cs

```diff
@@ -34,39 +33,53 @@ protected Saml2RedirectBinding BindInternal(Saml2Request saml2RequestResponse, s
         {
             base.BindInternal(saml2RequestResponse, signingCertificate);
 
-            var requestQueryString = string.Join("&", RequestQueryString(signingCertificate, messageName));
-            if (signingCertificate != null)
-            {
-                requestQueryString = SigneQueryString(requestQueryString, signingCertificate);
-            }
-
-            RedirectLocation = new Uri(string.Join("?", saml2RequestResponse.Destination.Uri.OriginalString, requestQueryString));
+            var uriBuilder = new UriBuilder(saml2RequestResponse.Destination.Uri.OriginalString);
+            NameValueCollection queryString = HttpUtility.ParseQueryString(uriBuilder.Query);
+            queryString.Add(RequestQueryString(signingCertificate, messageName));
+            uriBuilder.Query = queryString.ToString();
 
+            RedirectLocation = uriBuilder.Uri;
             return this;
         }
-        
+
         private string SigneQueryString(string queryString, X509Certificate2 signingCertificate)
         {
             var saml2Signed = new Saml2Sign(signingCertificate.PrivateKey);
             SignatureAlgorithm = signingCertificate.PrivateKey.SignatureAlgorithm;
             Signature = Convert.ToBase64String(saml2Signed.SignData(Encoding.UTF8.GetBytes(queryString)));
 
-            return string.Join("&", queryString, string.Join("=", Saml2Constants.Message.Signature, HttpUtility.UrlEncode(Signature)));
+            return HttpUtility.UrlEncode(Signature);
         }
 
-        private IEnumerable<string> RequestQueryString(X509Certificate2 signingCertificate, string messageName)
+        private NameValueCollection RequestQueryString(X509Certificate2 signingCertificate, string messageName)
         {
-            yield return string.Join("=", messageName, HttpUtility.UrlEncode(CompressRequest()));
+            var queryString = new NameValueCollection();
+            var message = HttpUtility.UrlEncode(CompressRequest());
+            if (!string.IsNullOrWhiteSpace(message))
+            {
+               queryString.Add(messageName, message);
+            }
 
             if (!string.IsNullOrWhiteSpace(RelayState))
             {
-                yield return string.Join("=", Saml2Constants.Message.RelayState, HttpUtility.UrlEncode(RelayState));
+                var relayState = HttpUtility.UrlEncode(RelayState);
+                if (relayState != null)
+                {
+                   queryString.Add(Saml2Constants.Message.RelayState, relayState);
+                }
             }
 
             if(signingCertificate != null)
             {
-                yield return string.Join("=", Saml2Constants.Message.SigAlg, HttpUtility.UrlEncode(signingCertificate.PrivateKey.SignatureAlgorithm));
+                var signatureAlgorithm = HttpUtility.UrlEncode(signingCertificate.PrivateKey.SignatureAlgorithm);
```

Comment:
> ReSharper seemed to think it could it return null, and it vigorously complained when this check was not included.

Nevermind, probably not going to make R# happy.